### PR TITLE
Fix path search

### LINKED ISSUE #46015: rustdoc: searches by path are broken
Current nightly:

image

Current stable:

image

cc @GuillaumeGomez

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
Fixes #46015.

r? @QuietMisdreavus

### CODE REVIEW

**Comment by QuietMisdreavus** [comment]
This changes the behavior when searching for paths, but it doesn't quite fix the issue?

<details><summary>Your PR</summary>

image

</details>

<details><summary>Current nightly</summary>

image

</details>

<details><summary>Current stable</summary>

image

</details>

**Comment by GuillaumeGomez** [comment]
Oh damn. Didn't take into account cases where the path doesn't start by the given search... I'll update it.

**Comment by GuillaumeGomez** [comment]
Ok fixed! The changed line is [this one](https://github.com/rust-lang/rust/pull/46081/files#diff-f3aa05ee448be83e65a501258d3aac2bR535).

**Comment by GuillaumeGomez** [comment]
After thinking about it once again (and a few tests), I found my solution not good enough so I strongly improved the `checkPath` function. Works way better now.

File: src/librustdoc/html/static/main.js [hunk 1]
```diff
@@ -374,11 +374,10 @@
...
                 split = valLower.split("::");
 
-            // remove empty keywords
-            for (var j = 0; j < split.length; ++j) {
-                split[j].toLowerCase();
-                if (split[j] === "") {
-                    split.splice(j, 1);
+            for (var z = 0; z < split.length; ++z) {
+                if (split[z] === "") {
+                    split.slice(z, 1);
```
Comment (Havvy) [inline src/librustdoc/html/static/main.js]: Slice returns a new array, so this is a no-op. Did you want to .splice() instead? From the removed code side, it looks like the answer is yes.

More idiomatically, you can also push those that raen't the empty string.

**Review by Havvy — CHANGES_REQUESTED** [review]

File: src/librustdoc/html/static/main.js (hunk 1 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]:
```
> x = [1,2,3];
(3) [1, 2, 3]
> x.splice(0, 1)
[1]
> x
(2) [2, 3]
```

So no, it clearly removes an entry from the array.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: Oh damn, you're absolutely right. Didn't see the missing 'p'!

**Comment by GuillaumeGomez** [comment]
Fixed.

**Review by Havvy — APPROVED** [review]
Just checking for basic JS stuff, and no other basic JS mistakes found.

File: src/librustdoc/html/static/main.js [hunk 2]
```diff
@@ -529,6 +526,49 @@
...
+
+                if (ty.parent && ty.parent.name) {
+                    path.push(ty.parent.name.toLowerCase());
+                }
+
+                if (startsWith.length > path.length) {
+                    return MAX_LEV_DISTANCE + 1;
+                }
+                for (var i = 0; i < path.length; ++i) {
+                    if (i + startsWith.length > path.length) {
```
Comment (Huntr0) [inline src/librustdoc/html/static/main.js]: You can use the condition used in the `if` statement as the `for` loop condition (when inverted), can't you?

File: src/librustdoc/html/static/main.js [hunk 3]
```diff
@@ -526,20 +526,47 @@
...
+                    var aborted = false;
+                    for (var x = 0; x < startsWith.length; ++x) {
+                        var lev = levenshtein(path[i + x], startsWith[x]);
+                        if (lev > MAX_LEV_DISTANCE) {
+                            aborted = true;
+                            break;
+                        }
+                        lev_total += lev;
+                    }
+                    if (aborted === false) {
```
Comment (Huntr0) [inline src/librustdoc/html/static/main.js]: Isn't it more common to check it like this: `!aborted`

File: src/librustdoc/html/static/main.js (hunk 2 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: Absolutely. I just have a personal preference to keep loop conditions as simple as possible. If someone else thinks it's a good idea, then I'll just update it.

File: src/librustdoc/html/static/main.js (hunk 3 above)
Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: It's slower. '!' tests (from memory): `length !== 0`, `0`, `null`, `undefined`. So clearly, it's way better in js to make the comparison like this.

Comment (Huntr0) [reply src/librustdoc/html/static/main.js]: Didn't know that. Thanks for explaining it.

Comment (Havvy) [reply src/librustdoc/html/static/main.js]: Except you can generally rely on the JIT to figure out that `aborted` can only contain a boolean.

Comment (GuillaumeGomez) [reply src/librustdoc/html/static/main.js]: I did too much rust, I *love* being explicit. 😛

**Comment by alexcrichton** [comment]
Looks like we forgot to backport this to 1.23.0 (sorry about that!) so removing beta tags

**Comment by alexcrichton** [comment]
Er sorry, looks like this was backported in https://github.com/rust-lang/rust/pull/46886

**Pull request merged**